src/azul/math.py

```diff
@@ -0,0 +1,24 @@
+def lossless_int(value: float) -> int:
+    """
+    Perform a lossless conversion of float to int. An AssertionError is raised
+    if the conversion would not be lossless.
+
+    >>> lossless_int(1.0)
+    1
+
+    >>> lossless_int(1.1)
+    Traceback (most recent call last):
+    AssertionError: 1.1
+
+    Note that values with a large count of significant figures can be subject
+    to rounding errors.
+
+    >>> lossless_int(100_000_000_000_000_000_000_000.0)
+    99999999999999991611392
```

Comment:
I'm afraid I didn't fully think this through:

`100_000_000_000_000_000_000_000.0`, which is the same as `1e23`, can't be represented exactly as a `float`:

```python
>>> from decimal import Decimal
>>> Decimal('1e23') - Decimal(1e23)
Decimal('8388608')
99999999999999991611392 + 8388608
100000000000000000000000
```

Adding 1 yields the next higher integer that can be represented as a `float`:

```python
>>> int(Decimal('100_000_000_000_000_000_000_001'))
100000000000000000000001
>>> int(float(Decimal('100_000_000_000_000_000_000_001')))
100000000000000008388608
```

When ES computes a sum or a count between [99999999999999991611392, 100000000000000000000000] that is probably going to be serialized to JSON as `1e23`. I think returning that in our response as `99999999999999991611392` would be misleading as it suggests a precision that isn't there.

Sorry for wavering on this but I think we should go back to returning all ES sums as floats.

Generally speaking, it appears that a decimal floating point literal is not a good way to represent a `float` type in textual form. It's not literal at all. A better literal representation would be a hexadecimal literal like what `float.hex` returns.

```python
>>> float(99999999999999991611392).hex()
'0x1.52d02c7e14af6p+76'
>>> float(99999999999999991611393).hex()
'0x1.52d02c7e14af6p+76'
>>> float(100000000000000000000000).hex()
'0x1.52d02c7e14af6p+76'
>>> float(100000000000000000000001).hex()
'0x1.52d02c7e14af7p+76'
```